README.rst
happycache: a utility for loading pages into cache
==================================================

.. image:: https://travis-ci.org/hashbrowncipher/happycache.svg?branch=master
    :target: https://travis-ci.org/hashbrowncipher/happycache

Databases such as PostgreSQL and Cassandra make extensive use of the OS page
cache to serve queries. Maintenance events like reboots result in the database
software operating against an empty or almost empty page cache, hampering
performance until the cache can fill.

happycache is a utility to quicky re-fill ths OS page cache using the pages
that were present previously. A happy cache makes for a happy database!

Command-line usage
------------------

Usage
~~~~~

::

  $ happycache
  Usage: happycache (dump|load) args...
    dump
      print out a map of pages that are currently in the page cache by recursively
      walking a directory.
        directory: the directory to walk
                     (default=current working directory)

    load [threads] [filename]
      load pages into the cache using a happycache dump file
        filename: the happycache dump file to read (default=stdin)
        threads: the number of threads to read with (default=32)

Examples
~~~~~~~~

happycache dump files contain the addresses (not the data) of the pages in the
cache. They are designed to be highly compressible, so we'll use gzip to cut
the map file down to size. In this example, we'll dump the page cache of a
Cassandra data directory.

::

  cassandra@box:/var/lib/cassandra$ happycache dump
  root@box:/var/lib/cassandra# reboot
  cassandra@box:/var/lib/cassandra# happycache load

We can use happycache load across servers. In the example below, we rsync files
from the `/var/lib/cassandra` directory on box1 to the
`/var/lib/cassandra/subdir` directory on box2. Because the happycache dump file
encodes relative paths, it has no problem working under the modified path.

::

  cassandra@box1:/var/lib/cassandra$ happycache dump
  root@box2:/var/lib/cassandra/subdir# rsync -a box1:/var/lib/cassandra/ .
  cassandra@box2:/var/lib/cassandra/subdir# happycache load

FAQ
---

Can I use the same happycache map on two different servers?
  Yep! If the two servers contain identical or similar files, then a happycache
  map from one server can be used to "copy" the page cache onto a second
  server. This is useful for Postgres replicas, where the files are identical
  or very similar.

Does a happycache dump file include the contents of pages?
  No. The happycache dump includes only enough information to *locate* pages.
  The dump does not contain any actual data from within the pages. As a result,
  happycache dumps are small, and do not leak any confidential data that may
  be in the page cache.

Does happycache loading have a performance impact?
  The purpose of happycache is to populate a page cache as quickly as the
  underlying hardware will allow. In cases where happycache is running beside
  another application (e.g. a database), happycache's reads will compete with
  I/O from that application. To mitigate this, happycache places itself in
  ``CLASS_IDLE`` I/O priority, meaning that the Linux CFQ disk scheduler will
  prioritize all other I/O above happycache's. This only works with the CFQ
  I/O scheduler, and NOT the noop or deadline schedulers.

Can happycache cause kernel panics (or plague, or lice)?
  The interfaces happycache uses to dump and fill page caches are quite
  boring and entirely non-invasive. The worst case scenario would be
  accidentally filling your page cache up with data that nobody has use for.

Should I run this after my nightly backups to repopulate the page-cache?
  Instead of using happycache to repopulate a busted page cache, I would
  consider using a backup tool which does not thrash the page cache in the
  first place. ``tar``, for example, works with the operating system to ensure
  that the files it reads do not remain in the page cache.

How can I use cgroups to preserve the page cache?
-------------------------------------------------

The need for happycache can be avoided in many cases by using cgroups to limit
the amount of page cache an application can consume. For instance, one can run
``mycommand`` in a memory cgroup with 50MB of memory like so::

  $ sudo cgcreate -g memory:backups -a $USER -t $USER
  $ cgset -r memory.limit_in_bytes=52428800 backups
  $ cgexec -g memory:backups mycommand


debian/changelog
happycache (0.1.2) unstable; urgency=medium

  * don't load more pages than the system has RAM

 -- Josh Snyder <joshs@netflix.com>  Tue, 30 Jul 2019 11:04:34 -0700

happycache (0.1.1) unstable; urgency=medium

  * do a single posix_fadvise pass

 -- Josh Snyder <joshs@netflix.com>  Mon, 29 Jul 2019 21:25:55 -0700

happycache (0.1.0) unstable; urgency=medium

  * performance improvement: switch to using fadvise for an initial pass

 -- Josh Snyder <joshs@netflix.com>  Thu, 08 Nov 2018 10:26:11 -0800

happycache (0.0.5) unstable; urgency=medium

  * bugfix: add a missing mutex unlock
  * bugfix: close the list correctly
  * bugfix: pop only one item per iteration

 -- Josh Snyder <josh@code406.com>  Thu, 08 Mar 2018 20:56:19 +0000

happycache (0.0.4) unstable; urgency=medium

  * more robust handling of list concurrency

 -- Josh Snyder <josh@code406.com>  Mon, 05 Mar 2018 19:28:25 +0000

happycache (0.0.3) unstable; urgency=medium

  * packaging bump

 -- Josh Snyder <josh@code406.com>  Fri, 26 Jan 2018 23:40:17 +0000

happycache (0.0.2) unstable; urgency=medium

  * XFS doesn't provide d_type, so we have to stat()

 -- Josh Snyder <josh@code406.com>  Wed, 10 Jan 2018 22:07:58 +0000

happycache (0.0.1) unstable; urgency=medium

  * Initial release.

 -- Josh Snyder <josh@code406.com>  Thu, 13 Oct 2016 00:10:07 -0700


list.c
#include <pthread.h>
#include <semaphore.h>
#include <stdatomic.h>
#include <stdbool.h>
#include <stdio.h>
#include <stddef.h>

#include "list.h"

sl * list_pop_head(sll * list) {
	sl * ret = NULL;

	pthread_mutex_lock(&list->head_lock);
	while(true) {
		ret = list->head;

		if(ret != NULL) {
			list->head = ret->next;
			if(list->head == NULL) {
					// If we have emptied the list, we need to reset the tail
					// pointer to equal head.
					pthread_mutex_lock(&list->tail_lock);
					// Double check. This could have changed.
					if(ret->next == NULL) {
						// Nobody has come along and modified the tail, so just
						// update the tail pointer and let the next enqueue
						// handle it.
						list->tail = &list->head;
					} else {
						// A concurrent writer has modified the tail (which is
						// ret->next). Just take what they did and put it in
						// list->head.
						list->head = ret->next;
					}

					pthread_mutex_unlock(&list->tail_lock);
			}
			break;
		} else if(list->closed) {
			break;
		} else {
			pthread_cond_wait(&list->cond, &list->head_lock);
		}
	}
	pthread_mutex_unlock(&list->head_lock);

	return ret;
}

void list_push_head(sll * list, sl * item) {
	bool locked_tail = false;

	pthread_mutex_lock(&list->head_lock);
	if(list->head == NULL) {
		pthread_mutex_lock(&list->tail_lock);
		locked_tail = true;
	}

	item->next = list->head;
	list->head = item;
	pthread_mutex_unlock(&list->head_lock);

	if(locked_tail) {
		pthread_mutex_unlock(&list->tail_lock);
	}

	pthread_cond_signal(&list->cond);
}

void list_push_tail(sll * list, sl * l) {
	bool lock_head = false;
	//Clear any existing value out
	l->next = NULL;

	pthread_mutex_lock(&list->tail_lock);
	if(list->tail == &list->head) {
		lock_head = true;
	}
	*list->tail = l;
	list->tail = &l->next;
	pthread_mutex_unlock(&list->tail_lock);

	if(lock_head) {
		// If we modified list->head, we need to lock head_lock to allow all
		// threads in the list_pop_head critical section to exit or begin
		// waiting on the condition variable.
		pthread_mutex_lock(&list->head_lock);
	}

	// We must do this unconditionally, because even if we didn't modify head,
	// there may be waiters stacked up from when the list was last empty.
	pthread_cond_signal(&list->cond);

	if(lock_head) {
		pthread_mutex_unlock(&list->head_lock);
	}
}

void list_init(sll * list) {
	list->head = NULL;
	list->tail = &list->head;
	list->closed = false;
	pthread_mutex_init(&list->head_lock, NULL);
	pthread_mutex_init(&list->tail_lock, NULL);
	pthread_cond_init(&list->cond, NULL);
}

void list_close(sll * list) {
	pthread_mutex_lock(&list->head_lock);
	list->closed = true;
	pthread_mutex_unlock(&list->head_lock);

	pthread_cond_broadcast(&list->cond);
}


list.h
#ifndef LIST_H_INCLUDED
#define LIST_H_INCLUDED

#include <pthread.h>
#include <semaphore.h>
#include <stdint.h>

#define container_of(ptr, type, member) ({			\
	const typeof( ((type *)0)->member ) *__mptr = (ptr);	\
	(type *)( (char *)__mptr - offsetof(type,member) );})

struct singly_linked {
	struct singly_linked * next;
};
typedef struct singly_linked sl;

struct singly_linked_list {
	sl * head;				//guarded by head_lock (RW)
	sl ** tail;				//guarded by tail_lock (RW)
	bool closed;			//guarded by head_lock (RW)
	pthread_cond_t cond;	//guarded by head_lock (waits)

	pthread_mutex_t head_lock;
	pthread_mutex_t tail_lock;
};
typedef struct singly_linked_list sll;

void list_init(sll * list);
sl * list_pop_head(sll * list);
void list_push_head(sll * list, sl * item);
void list_push_tail(sll * list, sl * l);
void list_close(sll * list);

#endif


dumping.c
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <stddef.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <zlib.h>

#include "dumping.h"

#define CHUNK_SIZE (1024 * 1024)

void dumper_init(
	struct dumper_state *state,
	gzFile outfile,
	sll * work_list
)
{
	state->work_list = work_list;
	state->outfile = outfile;
	state->open_directories = 1;
	pthread_mutex_init(&state->outfile_lock, NULL);
}

void dump_file(int fd,
	char* fullpath,
	gzFile outfile,
	pthread_mutex_t * outfile_lock
) {
	long page_size = sysconf(_SC_PAGESIZE);
	struct stat file_stat;
	if(fstat(fd, &file_stat) != 0) {
		fprintf(stderr, "%s: %s\n", strerror(errno), fullpath);
		return;
	}

	uint64_t file_size = file_stat.st_size;
	uint64_t file_pages = (file_stat.st_size + page_size - 1) / page_size;
	uint64_t processed_pages = 0;
	uint32_t chunk_pages = file_pages > CHUNK_SIZE ? CHUNK_SIZE : file_pages;

	unsigned char *mincore_vec = calloc(1, chunk_pages);
	if(mincore_vec == NULL) {
		fprintf(stderr, "%s: %s\n", strerror(errno), fullpath);
		return;
	}

	char* file_mmap = mmap(0, file_size, PROT_NONE, MAP_SHARED, fd, 0);
	while(file_pages > 0) {
		uint32_t chunk_pages = file_pages > CHUNK_SIZE ? CHUNK_SIZE : file_pages;
		uint64_t chunk_bytes = chunk_pages * page_size;

		if(file_mmap == MAP_FAILED) {
			fprintf(stderr, "%s: %s\n", strerror(errno), fullpath);
			goto out;
		}
		file_pages -= chunk_pages;

		int ret = mincore(file_mmap + processed_pages * 4096, chunk_bytes, mincore_vec);

		if(ret != 0) {
			fprintf(stderr, "mincore() %s: %s\n", strerror(errno), fullpath);
			goto out;
		}

		size_t i = 0;
		for(; i < chunk_pages; i++) {
			if(mincore_vec[i] & 0x01) {
				break;
			}
		}

		ssize_t last = -processed_pages;
		if(i < chunk_pages) {
			pthread_mutex_lock(outfile_lock);

			gzprintf(outfile, "%s\n", fullpath);
			for(; i < chunk_pages; i++) {
				if(mincore_vec[i] & 0x01) {
					gzprintf(outfile, "%lu\n", i - last);
					last = i;
				}
			}
			pthread_mutex_unlock(outfile_lock);
		}

		processed_pages += chunk_pages;
	}
	munmap(file_mmap, file_size);


out:
	free(mincore_vec);
}

void dump_dir(
	struct dir_info * current,
	struct dumper_state * state
) {
	struct dirent *ep;
	while((ep = readdir(current->dir))) {
		if(strcmp(ep->d_name, "..") == 0 ||
			strcmp(ep->d_name, ".") == 0) {
			continue;
		}

		int basename_len = strlen(ep->d_name);
		size_t concat_len = basename_len + current->len + 2;
		char * fullpath = malloc(concat_len);
		snprintf(fullpath, concat_len, "%s/%s", current->path, ep->d_name);

		int down_fd = openat(dirfd(current->dir), ep->d_name, O_RDONLY);
		if(down_fd == -1) {
			fprintf(stderr, "%s: %s\n", strerror(errno), fullpath);
			free(fullpath);
			continue;
		}

		unsigned char type = ep->d_type;

		struct stat buf;
		if(type == 0) {
			fstat(down_fd, &buf);
			switch(buf.st_mode & S_IFMT) {
				case S_IFDIR: type = DT_DIR; break;
				case S_IFREG: type = DT_REG; break;
			}
		}

		if(type & DT_DIR) {
			atomic_fetch_add(&state->open_directories, 1);
			//Push the context we were working on to the shared stack
			list_push_head(state->work_list, &current->list);

			//Create a new context
			current = malloc(sizeof(struct dir_info));
			if(current == NULL) {
				perror(NULL);
				exit(1);
			}

			current->path = fullpath;
			current->dir = fdopendir(down_fd);
			current->len = concat_len;
		} else {
			if(type & DT_REG) {
				dump_file(
					down_fd,
					fullpath,
					state->outfile,
					&state->outfile_lock
				);
			}
			free(fullpath);
			close(down_fd);
		}
	}

	uint32_t ret = atomic_fetch_add(&state->open_directories, -1);
	if(ret == 1) {
		list_close(state->work_list);
	}

	//We just exhausted the directory
	closedir(current->dir);
	free(current->path);
	free(current);
}

void dump_worker(struct dumper_state * state) {
	struct dir_info * work;
	while(true) {
		work = container_of(
			list_pop_head(state->work_list),
			struct dir_info,
			list
		);
		if(work == NULL) {
			break;
		}
		dump_dir(
			work,
			state
		);
	}
}


happycache.c
#define _GNU_SOURCE
#define _XOPEN_SOURCE 700

#include <dirent.h>
#include <fcntl.h>
#include <pthread.h>
#include <semaphore.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifdef __linux__
#include <sys/sysinfo.h>
#endif
#include <unistd.h>
#include <zlib.h>

#define CHUNK_SIZE (1024 * 1024)
#define MAP_FILENAME ".happycache.gz"

#include "dumping.h"
#include "list.h"

long page_size;
int page_shift;

struct read_work {
	struct singly_linked list;
	int fd;
	void * addr;
	void * base_addr;
};
typedef struct read_work read_work;

struct fd_info {
	void * base_addr;
	size_t num_pages;
	int64_t refcount;
	unsigned char * mincore;
	size_t mincore_start;
};
typedef struct fd_info fd_info;

void finished_file_op(fd_info * fdi, int fd) {
	if(--fdi->refcount == 0) {
		if(fdi->base_addr != MAP_FAILED) {
			munmap(fdi->base_addr, fdi->num_pages << page_shift);
		}
		close(fd);
	}
}

struct loader_state {
	sll * work_list;
	sll * free_list;
};

uint16_t get_concurrency() {
	/*
	 * Ideally this would map directly to the maximum queue depth of the
	 * underlying block device. But that's impossible. So instead we multiply
	 * the number of CPUs by 8. Fun!
	 */
	return sysconf(_SC_NPROCESSORS_ONLN) * 8;
}

void read_worker(struct loader_state * state) {
	volatile long val = 0;
	while(true) {
		sl * item = list_pop_head(state->work_list);
		if(item == NULL) {
			break;
		}
		read_work * my_work = container_of(
			item,
			read_work,
			list
		);
		val ^= *(long *)(my_work->addr);
		list_push_tail(state->free_list, &my_work->list);
	}
}

int prepare_file(char * line, fd_info * fdi) {
	int fd = open(line, O_RDONLY | O_CLOEXEC);
	if(-1 == fd) {
		fprintf(stderr, "Could not open %s: ", line);
		perror(NULL);
		return -1;
	}

	fdi->base_addr = MAP_FAILED;
	fdi->refcount = 1;
	fdi->num_pages = -1;
	fdi->mincore_start = 0;

	struct stat file_stat;
	if(fstat(fd, &file_stat) != 0) {
		fprintf(stderr, "Could not fstat %s: ", line);
		perror(NULL);
		return fd;
	}

	uint64_t num_pages = (file_stat.st_size + page_size - 1) / page_size;

	fdi->base_addr = mmap(0, num_pages << page_shift, PROT_READ, MAP_SHARED, fd, 0);
	if(fdi->base_addr == MAP_FAILED) {
		fprintf(stderr, "Could not mmap %s: ", line);
		perror(NULL);
		return fd;
	}

	//Only update the struct's num_pages once we're confident everything's
	//tip-top
	fdi->num_pages = num_pages;

	uint32_t chunk_pages = num_pages > CHUNK_SIZE ? CHUNK_SIZE : num_pages;
	fdi->mincore = malloc(chunk_pages);
	mincore(fdi->base_addr, chunk_pages * page_size, fdi->mincore);
	return fd;
}

void load_pages(
	fd_info * fdi
	, int fd
	, uint64_t start
	, uint64_t count
) {
	if(fd == -1) {
		return;
	}

	uint64_t num_pages = fdi->num_pages;

	if(num_pages - start < count) {
		count = num_pages - start;
	}

	while(count > 0) {
		uint64_t mincore_offset = start - fdi->mincore_start;
		if(mincore_offset > CHUNK_SIZE) {
			fdi->mincore_start = start;
			mincore_offset = 0;
			uint64_t remaining = num_pages - start;
			uint64_t chunk_pages = remaining > CHUNK_SIZE ? CHUNK_SIZE : remaining;

			mincore(
				fdi->base_addr + fdi->mincore_start * page_size,
				chunk_pages * page_size,
				fdi->mincore
			);
		}

		if((fdi->mincore[mincore_offset] & 0x01) == 0) {
			// 8 pages (32KB) is much smaller than the 128KB typical max_sectors_kb
			uint8_t limit = 8;
			if(count < limit) {
				limit = count;
			}

			posix_fadvise(fd, start << page_shift, limit << page_shift, POSIX_FADV_WILLNEED);
			count -= limit;
			start += limit;
		} else {
			start += 1;
			count -= 1;
		}
	}
}

int load_from_map(gzFile map) {
	page_size = sysconf(_SC_PAGESIZE);
	page_shift = __builtin_ctz(page_size);

#ifdef __linux__
	struct sysinfo info;
	sysinfo(&info);
	uint64_t pages_left = info.totalram >> page_shift;
#else
	uint64_t pages_left = -1;
#endif


	fd_info fdi;
	bzero(&fdi, sizeof(fd_info));

	char line[4096];
	int fd = -1;
	//page == -1 signifies that we are between files
	int64_t page = -1;
	uint64_t count = 0;

	while(pages_left--) {
		if(gzgets(map, line, sizeof(line)) == NULL) {
			if(!gzeof(map)) {
				//The file is corrupt?
				fputs(gzerror(map, NULL), stderr);

				return 1;
			}

			load_pages(&fdi, fd, page - count, count);
			break;
		}

		size_t len = strlen(line) - 1;
		//Trim the trailing newline.
		line[len] = 0;

		if(page >= 0) {
			char * endptr;
			unsigned long skip = strtoul(line, &endptr, 10);
			if(endptr - line != len) {
				//We can't parse this as a number, so it must be a filename.
				//Filenames start either with ./ or /, so this works.
				if(fd != -1) {
					load_pages(&fdi, fd, page - count, count);
					finished_file_op(&fdi, fd);
				}
				page = -1;
			} else if(skip <= 1) {
				page += skip;
				count += 1;
			} else {
				load_pages(&fdi, fd, page - count, count);
				page += skip;
				count = 1;
			}
		}

		if(page < 0) {
			fd = prepare_file(line, &fdi);

			// This is awful. It makes way more sense as page = 0
			page = 1;
			count = 0;
		}
	}

	return 0;
}

void do_usage(char* name) {
	fprintf(stderr, "Usage: %s (dump|load) args...\n", name);
	fprintf(stderr, "  dump\n");
	fprintf(stderr, "    save a map of pages that are currently in the page cache by recursively\n");
	fprintf(stderr, "    walking the current working directory.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  load [threads] [filename]\n");
	fprintf(stderr, "    load pages into the cache using a happycache dump file\n");
	fprintf(stderr, "      threads: the number of threads to read with (default=32)\n");
	fprintf(stderr, "      filename: the happycache dump file to read (default=.happycache.gz)\n");

	exit(1);
}

void do_load(int argc, char** argv, char * progname) {
	if(argc > 2) {
		do_usage(progname);
	}

	char * to_read = MAP_FILENAME;
	if(argc >= 2) {
		to_read = argv[1];
	}

	gzFile map = NULL;
	map = gzopen(to_read, "rb");
	if(map == NULL) {
		perror("Could not open map file");
		exit(1);
	}

	int ret = load_from_map(map);
	gzclose(map);

	if(ret) {
		exit(ret);
	}

	exit(0);
}

void do_dump(int argc, char ** argv, char * progname) {
	if(argc > 1) {
		do_usage(progname);
	}

	char * outfile_name = alloca(32);
	strncpy(outfile_name, ".happycache.gz.XXXXXX", 32);
	int outfile_fd = mkostemp(outfile_name, O_CLOEXEC);
	if(-1 == outfile_fd) {
		fprintf(stderr, "Could not create temporary output file: ");
		perror(NULL);
		exit(1);
	}

	gzFile outfile = gzdopen(outfile_fd, "wb1");

	struct dir_info * work = malloc(sizeof(struct dir_info));
	char * filename = ".";
	work->dir = opendir(filename);
	if(NULL == work->dir) {
		fprintf(stderr, "Could not open directory %s: ", filename);
		perror(NULL);
		exit(1);
	}

	sll work_list;
	list_init(&work_list);

	struct dumper_state state;
	dumper_init(
		&state,
		outfile,
		&work_list
	);

	work->len = 1;
	work->path = malloc(work->len + 1);
	strncpy(work->path, filename, work->len + 1);

	int num_threads = get_concurrency();
	pthread_t * threads = calloc(num_threads, sizeof(pthread_t));
	for(uint32_t i = 0; i < num_threads; i++) {
		pthread_create(
			&threads[i],
			NULL,
			(void * (*)(void *)) dump_worker,
			(void *) &state
		);
	}

	list_push_head(&work_list, &work->list);

	for(uint32_t i = 0; i < num_threads; i++) {
		pthread_join(threads[i], NULL);
	}

	gzclose(outfile);
	rename(outfile_name, MAP_FILENAME);
}

int main(int argc, char** argv) {
	if(argc < 2) {
		do_usage(argv[0]);
	}

	struct sched_param param;
	param.sched_priority = 0;
	sched_setscheduler(0, SCHED_IDLE, &param);

	if(strcmp(argv[1], "load") == 0) {
		do_load(argc - 2, &argv[2], argv[0]);
	} else if(strcmp(argv[1], "dump") == 0) {
		do_dump(argc - 2, &argv[2], argv[0]);
	} else {
		do_usage(argv[0]);
	}
}
